### PR TITLE
File synchronization

### PR DESCRIPTION
**Pull request by Einh06**
This is a work in progress PR for file synchonization.

- Small renaming and model changes for file. 
- Changed path processing from `String` based to `URL` (much more reliable
- Creation of a base file structure for displaying with a root, then the 3 main folders: My data, Courses and Shared
- Fetching shared data from different the `/files`

### CODE REVIEW

File: schulcloud/FileHelper.swift [hunk 1]
```diff
@@ -12,11 +12,187 @@ import BrightFutures
 import CoreData
 import Marshal
 
-class FileHelper {
+class FileSync : NSObject {
```
Comment (mathebox) [inline schulcloud/FileHelper.swift]: Is there a reason to inherit from NSObject?

File: schulcloud/FileHelper.swift [hunk 2]
```diff
@@ -12,11 +12,187 @@ import BrightFutures
...
+
+    typealias ProgressHandler = (Float) -> ()
+    
+    fileprivate var fileTransferSession : URLSession! = nil
```
Comment (mathebox) [inline schulcloud/FileHelper.swift]: You can remove `= nil`

File: schulcloud/FileHelper.swift [hunk 3]
```diff
@@ -12,11 +12,187 @@ import BrightFutures
...
+    fileprivate let fileDataSession : URLSession
+
+    var runningTask : [Int: Promise<Data, SCError>] = [:]
+    var progressHandlers : [Int : ProgressHandler] = [:]
+
+    override init() {
+
```
Comment (mathebox) [inline schulcloud/FileHelper.swift]: whitespace

File: schulcloud/FileHelper.swift [hunk 4]
```diff
@@ -12,11 +12,187 @@ import BrightFutures
...
+    }
+    
+    private func getUrl(for file: File) -> URL? {
+        var urlComponent = URLComponents(url: fileStorageURL, resolvingAgainstBaseURL: false)!
+        urlComponent.query = "path=\(file.url.absoluteString)"
+        return try? urlComponent.asURL()
+    }
+    
+    private func request(for url: URL) -> URLRequest {
+
```
Comment (mathebox) [inline schulcloud/FileHelper.swift]: whitespace

File: schulcloud/FileHelper.swift [hunk 5]
```diff
@@ -12,11 +12,187 @@ import BrightFutures
...
+            } catch let error as SCError {
+                promise.failure(error)
+            } catch let error {
+                promise.failure(.jsonDeserialization(error.localizedDescription) )
+            }
+        }.resume()
+        return promise.future
+    }
+
+    func sharedDownload() -> Future<[[String:Any]], SCError> {
```
Comment (mathebox) [inline schulcloud/FileHelper.swift]: `downloadSharedFiles` might be a better naming

File: schulcloud/LoadingViewController.swift [hunk 6]
```diff
@@ -64,6 +65,30 @@ class LoadingViewController: UIViewController  {
     }
     
     func startDownload() {
+
```
Comment (mathebox) [inline schulcloud/LoadingViewController.swift]: whitespace

File: schulcloud/LoadingViewController.swift [hunk 7]
```diff
@@ -92,7 +117,7 @@ class LoadingViewController: UIViewController  {
             .onFailure { error in
                 self.show(error: error)
         }
-        
+       */
```
Comment (mathebox) [inline schulcloud/LoadingViewController.swift]: Can't this block be removed?

File: schulcloud/FilesViewController.swift [hunk 8]
```diff
@@ -8,10 +8,12 @@
 
 import UIKit
 import CoreData
+import BrightFutures
 
 class FilesViewController: UITableViewController, NSFetchedResultsControllerDelegate {
 
     var currentFolder: File!
+    var fileSync = FileSync()
```
Comment (Einh06) [inline schulcloud/FilesViewController.swift:16]: I'm creating `URLSession`s for each controller, I use that to be able to cancel tasks when this object is destroyed

File: schulcloud/FilesViewController.swift [hunk 9]
```diff
@@ -33,16 +34,58 @@ class FilesViewController: UITableViewController, NSFetchedResultsControllerDele
...
+            .map { json -> () in
+                FileHelper.updateDatabase(contentsOf: self.currentFolder, using: json)
+                return ()
+            }.onSuccess(callback: successBlock)
+            .onFailure { (error) in
+                log.error(error)
+            }.onComplete { (_) in
+                completeBlock()
+            }
+        }
```
Comment (mathebox) [inline schulcloud/FilesViewController.swift]: DRY. You have the calls `onSuccess`, `onFailure` and `onComplete` three times. Just create a variable for the future before the if/elseif/else block and do the future handling once afterwards.

**Review by mathebox — CHANGES_REQUESTED** [review]
Please also use the CoreData notification to update the course files folders.

~~When I'm testing, no files show up. I'm getting an `SCError.apiError`~~
It appears that is was network issue on my side.

File: schulcloud/FileHelper.swift [hunk 10]
```diff
@@ -141,33 +127,37 @@ class FileSync : NSObject {
...
+            } catch let error as SCError {
+                promise.failure(error)
             } catch let error {
                 promise.failure(.jsonDeserialization(error.localizedDescription) )
             }
         }.resume()
         return promise.future
     }
+    
+    private func confirmNetworkResponse(data: Data?, response: URLResponse?, error: Error?) throws -> Data{
```
Comment (mathebox) [inline schulcloud/FileHelper.swift]: Instead of `throws` I would like to see a `Result<>` returned

File: schulcloud/FileHelper.swift [hunk 11]
```diff
@@ -141,33 +127,37 @@ class FileSync : NSObject {
...
+        guard error == nil else {
```
Comment (mathebox) [inline schulcloud/FileHelper.swift]: use `if let` to unpack to error

**Review by mathebox — CHANGES_REQUESTED** [review]

File: schulcloud/FilesViewController.swift (hunk 9 above)
Comment (Einh06) [reply schulcloud/FilesViewController.swift]: Each Futures are of different type. I will see if I can factor it, but unsure if it will work

Comment (Einh06) [reply schulcloud/FilesViewController.swift]: I have experimented with some solution for fixing that, maping all Futures to a `Future<Any, SCError>` to bind them to a variable and then chain the success/failure/complete. I think DRY does not make things more readable, complicate things for no value. As much as i agree that repeating feels wrong, it is a superior solution than any alternative in this context.

Comment (Einh06) [reply schulcloud/FilesViewController.swift]: I would add that the meat of the problem isn't the repeating part, more that we have 2 special cases to handle, for courses directories and shared directory. Due to current requirement, this is unavoidable.

**Review by mathebox — APPROVED** [review]

**Pull request merged**